dl/src/main/scala/com/intel/analytics/bigdl/nn/TimeDistributedCriterion.scala

```diff
@@ -29,77 +29,66 @@ import scala.reflect.ClassTag
  * @param critrn
  */
 
-class TimeDistributedCriterion[T : ClassTag](critrn : TensorCriterion[T])
+class TimeDistributedCriterion[T : ClassTag](critrn : TensorCriterion[T],
+                                             sizeAverage: Boolean = true)
 (implicit ev: TensorNumeric[T]) extends TensorCriterion[T] {
 
-  private val fInput: Tensor[T] = Tensor[T]()
-  private val fTarget: Tensor[T] = Tensor[T]()
-  private var inputSize: Array[Int] = _
-  private var targetSize: Array[Int] = _
-
-  private def combine(src: Array[Int], target: Array[Int]): Unit = {
-    require(src.length == target.length + 1,
-      "TimeDistributedCriterion: combine method requires src.length == target.length + 1" +
-        s" Current src.length = ${src.length}" +
-        s" Current target.length = ${target.length}")
-    target(0) = src(0) * src(1)
-    var j = 1
-    while (j < target.length) {
-      target(j) = src(j + 1)
-      j += 1
-    }
-  }
+  private var fInput: Tensor[T] = Tensor[T]()
+  private var fTarget: Tensor[T] = Tensor[T]()
 
   override def updateOutput(input: Tensor[T], target: Tensor[T]): T = {
     /**
+     * Take each time slice of input and target, and add up all outputs of slices
      * For example
-     * input.size = [B, T, D] => inputSize = [B * T, D]
-     * target.size = [B, T] => targetSize = [B * T]
+     * input.size = [B, T, D] => fInput.size = [B, D]
+     * target.size = [B, T] => fTarget.size = [B]
+     * If sizeAverage is true, the output is averaged through time dimension
      */
-    if (inputSize == null) {
-      inputSize = new Array[Int](input.size.length - 1)
+    val nstep = input.size(2)
+    var i = 1
+    while (i <= nstep) {
+      fInput = input.select(2, i)
+      fTarget = target.select(2, i)
+      output = ev.plus(output, critrn.updateOutput(fInput, fTarget))
+      i += 1
     }
-    if (targetSize == null) {
-      targetSize = new Array[Int](target.size.length - 1)
+    if (sizeAverage) {
+      output = ev.divide(output, ev.fromType[Int](nstep))
     }
-
-    combine(input.size, inputSize)
-    combine(target.size, targetSize)
-    fInput.set(input).resize(inputSize)
-    fTarget.set(target).resize(targetSize)
-    output = critrn.updateOutput(fInput, fTarget)
     output
   }
 
   override def updateGradInput(input: Tensor[T], target: Tensor[T]): Tensor[T] = {
-    val _gradInput = critrn.updateGradInput(fInput, fTarget).toTensor[T]
-    gradInput = _gradInput.resize(input.size)
+    /**
+      * Take each time slice of input and target, and calculate gradInput of each slice
+      * If sizeAverage is true, the gradInput is also averaged through time dimension
+      */
+    gradInput.resizeAs(input)
+    gradInput.zero()
+    var _gradInput = Tensor[T]()
```

Comment:
define _gradInput as global variable like fInput and fTarget. Use val instead of var